File: src/GnOuGo.Flow.Core/Runtime/Executors/LoopParallelExecutor.cs
using System.Text.Json.Nodes;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Parallel loop: iterates over items in parallel.
/// </summary>
public sealed class LoopParallelExecutor : IStepExecutor
{
    public string StepType => "loop.parallel";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The step is missing nested `steps`, its input is not an object, or `items` is not an array."),
        new(ErrorCodes.LoopLimit, false, "The number of loop items exceeds the runtime loop limit.")
    };

    public string DslSnippet => """
        ### loop.parallel — Iterate items in parallel
        ```yaml
        - id: process
          type: loop.parallel
          input:
            items: "${data.inputs.my_list}"    # required — array expression
            max_concurrency: 2                  # optional
          item_var: item                        # variable name for current item (default: "item")
          index_var: idx                        # variable name for current index (default: "i")
          steps:
            - id: transform
              type: template.render
              input: { engine: mustache, template: "{{val}}", data: { val: "${data.item}" }, mode: text }
        ```
        Context: `data.<item_var>` (current item), `data.<index_var>` (current index).
        Output: `{ results: [...], count: N }`
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var subSteps = ctx.Step.Steps
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "loop.parallel requires 'steps'");

        var input = ctx.Engine.GetResolvedInput(ctx);
        var inputObj = input as JsonObject
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "loop.parallel input must be object");

        if (!inputObj.TryGetPropertyValue("items", out var itemsNode) || itemsNode is not JsonArray items)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "loop.parallel requires 'items' array");

        var itemVar = ctx.Step.Source.ItemVar ?? "item";
        var indexVar = ctx.Step.Source.IndexVar ?? "i";

        int maxConcurrency = 0;
        if (inputObj.TryGetPropertyValue("max_concurrency", out var mc) && mc != null)
            maxConcurrency = (int)ExpressionEvaluator.GetNumber(mc);

        if (items.Count > ctx.Limits.MaxLoopIterations)
            throw new WorkflowRuntimeException(ErrorCodes.LoopLimit,
                $"Loop items ({items.Count}) exceeds limit ({ctx.Limits.MaxLoopIterations})");

        var semaphore = maxConcurrency > 0 ? new SemaphoreSlim(maxConcurrency) : null;
        var tasks = new List<Task<(int index, JsonObject iterData)>>();

        for (int i = 0; i < items.Count; i++)
        {
            var index = i;
            var item = items[i];

            tasks.Add(Task.Run<(int, JsonObject)>(async () =>
            {
                if (semaphore != null) await semaphore.WaitAsync(ct);
                try
                {
                    var iterData = ctx.Data.DeepClone() as JsonObject ?? new JsonObject();
                    iterData[itemVar] = item?.DeepClone();
                    iterData[indexVar] = JsonValue.Create(index);
                    iterData["_loop"] = new JsonObject { ["index"] = JsonValue.Create(index), ["item"] = item?.DeepClone() };
                    iterData["loop"] = new JsonObject { ["index"] = JsonValue.Create(index), ["item"] = item?.DeepClone() };

                    var iterResult = new RunResult { Success = true };
                    await ctx.Engine.ExecuteStepsAsync(subSteps, iterData, iterResult, ctx.Limits, ctx.CallDepth, ctx.CallStack, ct, ctx.TelemetrySpan);
                    return (index, iterData);
                }
                finally
                {
                    semaphore?.Release();
                }
            }, ct));
        }


        var results = await Task.WhenAll(tasks);

        var output = new JsonArray();
        foreach (var (_, iterData) in results.OrderBy(r => r.index))
        {
            var stepData = iterData["steps"]?.DeepClone() as JsonObject ?? new JsonObject();
            // Clean up temp keys
            var keysToRemove = stepData.Select(kv => kv.Key).Where(k => k.StartsWith("__") && k.EndsWith("__")).ToList();
            foreach (var key in keysToRemove) stepData.Remove(key);
            output.Add((JsonNode)stepData);
        }

        return new JsonObject { ["results"] = output, ["count"] = items.Count };
    }
}

File: clients/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/loop_parallel_executor.py
from __future__ import annotations

from gnougo_flow_core.runtime import *  # noqa: F401,F403

class LoopParallelExecutor:
    step_type = "loop.parallel"
    step_description = "Iterate items in parallel with optional max_concurrency."
    dsl_snippet = """
### loop.parallel - Execute iterations in parallel over items
```yaml
- id: fanout
  type: loop.parallel
  item_var: item
  index_var: i
  input:
    items: "${data.inputs.urls}"
    max_concurrency: 4
  steps:
    - id: process
      type: set
      input:
        index: "${i}"
        value: "${item}"
```
Output: `{ results: [<iteration_steps>...], count: <number> }`.
"""
    documented_exceptions = [
        (ErrorCodes.INPUT_VALIDATION, False, "loop.parallel requires object input and 'items' array."),
        (ErrorCodes.LOOP_LIMIT, False, "loop.parallel exceeded items limit."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        if not ctx.step.steps:
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "loop.parallel requires 'steps'")
        input_obj = ctx.engine.get_resolved_input(ctx)
        if not isinstance(input_obj, dict):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "loop.parallel input must be object")

        items = input_obj.get("items")
        if not isinstance(items, list):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "loop.parallel requires 'items' array")
        if len(items) > ctx.limits.max_loop_iterations:
            raise WorkflowRuntimeException(ErrorCodes.LOOP_LIMIT, "Loop items exceed limit")

        item_var = ctx.step.source.item_var or "item"
        index_var = ctx.step.source.index_var or "i"
        max_concurrency = int(input_obj.get("max_concurrency", 0))
        sem = asyncio.Semaphore(max_concurrency) if max_concurrency > 0 else None

        async def run_iter(index: int, item: Any) -> tuple[int, dict[str, Any]]:
            if sem:
                await sem.acquire()
            try:
                iter_data = {
                    "inputs": dict(ctx.data.get("inputs", {})),
                    "steps": dict(ctx.data.get("steps", {})),
                    "env": dict(ctx.data.get("env", {})),
                    item_var: item,
                    index_var: index,
                    "_loop": {"index": index, "item": item},
                    "loop": {"index": index, "item": item},
                }
                run = RunResult(success=True)
                await ctx.engine.execute_steps_async(ctx.step.steps, iter_data, run, ctx.limits, ctx.call_depth, set(ctx.call_stack), ctx.telemetry_span)
                return index, iter_data
            finally:
                if sem:
                    sem.release()

        results = await asyncio.gather(*(run_iter(i, it) for i, it in enumerate(items)))
        ordered = [r for _, r in sorted(results, key=lambda x: x[0])]
        cleaned = []
        for item in ordered:
            steps = {k: v for k, v in item.get("steps", {}).items() if not (k.startswith("__") and k.endswith("__"))}
            cleaned.append(steps)
        return {"results": cleaned, "count": len(items)}
